### ISSUE TITLE
Consistency problem between documents assigned to topics and topics name

### ISSUE THREAD

**Issue by ruVdi** [MaartenGr/BERTopic#1339]
Hi Marteen and hi to the whole community. 
I am writing because I am trying to use BERTopic to analyse about 12,000 maintenance-related documents. First of all I pre-processed the data using algorithms for removing stopwords and punctuation. Then I used BERTopic and of 12000 documents about 4800 documents (~40% of the total) are considered outliers. By using outlier reduction techniques, I manage to reduce these considerably although I do not want to force this step too much to avoid increasing the noise within the generated topics. After training the BERTopic model, I realised that the documents that are assigned to a specific topic are not consistent with the words in the "Name" column generated with the .get_document_info(docs) or .get_topic_info(docs) commands. In addition to many irrelevant documents, many documents that should belong to other topics in the list (still referring to the representative words in the "Name" column) are also assigned to the wrong topic. I have tried many different settings both in BERTopic's internal models (CountVectorizer, SentenceTransformer, KeyBERTInspired, ClassTfidfTransformer, UMAP, HDBSCAN) and within BERTopic, but this problem keeps recurring and I can't understand why. 
I have seen that in BERTopic it is not possible to force the assignment of a document to a specific topic, so I would like to find out how I can fix this error. Furthermore, as an alternative to HDBSCAN, I have tried using K-Means, which does not produce any outliers, but I am still not satisfied with the results as they contain a lot of noise.
Do you have any advice on what parameters I should change or what route I should take to improve the assignment of documents to topics generated with BERTopic in order to get consistent topics?

I will try to better explain the problem I am experiencing. By manually analyzing the documents assigned to each topic, I realized that many documents regarding, for example, "Password request for machinery X" are assigned to a topic that is about "Oil leakage of machinery Y" even if BERTopic has generated a topic dedicated to "Password request for machinery X" and so on for almost all the generated topics. In fact, in each topic, there are documents that should belong to another topic as well as a consistent number of outlier documents that have nothing to do with the representative words in the "Name" column, and I cannot figure out which parameter/model/algorithm is responsible for this error. In other words, I am experiencing a consistency problem between the representative words in the "Name" column obtained with .get_topic_info() or .get_documents_info() and the documents that are actually assigned to the topics by BERTopic.
If it helps, I used n_gram_range=(2,6) in the CountVectorizer and in BERTopic and set nr_topics='auto'. 

I recognize the huge potential of this algorithm and would be happy to have a discussion with you on it so that I can understand more about it. I remain at disposal for any clarifications regarding the value of the parameters used within the algorithm.

Thanks!

**Comment by ruVdi** [comment]
First of all, thank you for your quick response. Yes, it seems to be like that, the representative words of each topic reported in the "Name" column are correct in fact all the topics created are about problems that actually occur many times in the 12000 documents. The problem is that if I manually inspect the content of the topics, only a portion of the documents are actually about "Name" (obtained with .get_topic_info()) while many documents are noise (may be due to the fact that I use .reduce_outliers() to redistribute the outliers) and another portion of documents should belong to other topics considering the "Name" column obtained with the .get_topic_info() command.

This is the code I use to train BERTopic:
```python
vectorizer_model = CountVectorizer(ngram_range=(2,6), stop_words=None, min_df=0.05)

sentence_model = SentenceTransformer("bert-base-uncased")

representation_model = KeyBERTInspired(top_n_words=15, nr_repr_docs=12)

ctfidf_model = ClassTfidfTransformer(bm25_weighting=True, reduce_frequent_words=False)

umap_model = UMAP(n_neighbors=25, n_components=2, metric='cosine', min_dist=0.1, random_state=42, low_memory=False)

hdbscan_model = HDBSCAN(min_cluster_size=15, min_samples=10, metric='manhattan', cluster_selection_method='eom', gen_min_span_tree=True, prediction_data=True)

bertopic_model = BERTopic(
                    vectorizer_model=vectorizer_model,
                    embedding_model=sentence_model,
                    umap_model = umap_model,
                    hdbscan_model = hdbscan_model,
                    representation_model=representation_model,
                    ctfidf_model=ctfidf_model,
                    nr_topics='auto',
                    language="english",
                    low_memory=False,
                    verbose=True, 
                    calculate_probabilities=True, 
                    top_n_words=15,
                    n_gram_range=(2,6))

topics, probs = bertopic_model.fit_transform(docs, embeddings)

topics = bertopic_model.reduce_outliers(docs, topics, strategy="c-tf-idf", threshold=0.6)

bertopic_model.update_topics(docs, topics=topics)
```
Once I have calculated topics and probs with the .fit_transform() command and reduced the outliers, I run the following code to create a new column in my original excel dataframe where I associate the Topic ID to each document:

```python
topic_info = bertopic_model.get_document_info(docs)

topics_1 = topic_info['Topic']

df['Topic'] = topics_1
```

At this point, if I go to manually inspect the content of a specific topic, I notice this inconsistency between the documents contained in the topic and what should actually be the main focus of the topic.

**Comment by MaartenGr** [comment]
I forgot to ask, which version of BERTopic are you using? 

There are a few things in your code that I think could be optimized which might explain your results. First, the embedding model. You will need a model that can generate accurate semantic representations of text but a "bert-base-uncased" model typically does not. This is because those embeddings are not optimized for semantic similarity which is important for clustering tasks. Instead, if you would use "all-mpnet-base-v2" then the performance should increase drastically. Second, by setting `nr_components=2`, you are removing a lot of information from the original embeddings thereby decreasing its representative power. Setting that to 5 typically generates much better results. Third, I am not sure whether the "manhattan" distance would be more accurate than euclidean since I do not believe the data necessarily follows such a pattern. Fourth, it might be worthwhile to check which documents are misaligned and whether those were assigned through the reduce outliers function. That helps you understand whether the misalignment is a result of the initial training or through outlier reduction.

**Comment by ruVdi** [comment]
Thanks for the advice, I am using Version: 0.15.0. I tried setting the parameters differently and in a few aspects the results of BERTopic improved. I have noticed a considerable reduction in the number of outliers (there are now around 2800 compared to 4800), but still if I inspect the documents contained in a specific topic in detail I notice this problem of inconsistency between the representative words of a topic and the documents that are actually assigned to that topic. I have performed many tests using different parameters but each run presents this problem. What I have noticed is that for topics where BERTopic assigns up to 100 documents, the content is consistent with what the representative words describe, but for more populated topics the content is inconsistent.
For example, topic number 2 is assigned 755 documents and of these only 310 actually relate to what the topic is supposed to mean, while the remaining documents should belong to other topics or have nothing to do with it.Furthermore, what I have noticed is that if a cell in my excel file (i.e. a single document) contains a sentence such as 'Password request for machinery X for engine breakage', this document is assigned to a topic concerning 'engine breakage' rather than the topic 'Password request for machinery X'.
Another problem I have noticed is that if a cell in my excel file (so a single document) contains a sentence such as "Password request for machinery X for engine breakage", this document is assigned to a topic concerning "engine breakage" rather than to the topic "Password request for machinery X". I point this out because in the topic "Password request for machinery X" there should be around 3500 documents but at most BERTopic assigns 600 documents. This may not be a mistake because the document is assigned to a topic concerning engine breakage, but this topic is much less relevant than the topic "Password request for machinery X".
I'm sure I'm doing something wrong, maybe even in the data I provide as input to BERTopic. The 12000 documents are contained in 12000 cells in one column of an excel file. The data are without stopwords and punctuation, are in lower case and have been tokenized and lemmatised using SpaCy. Then I transform these documents into a list using the command
```
docs= list(df['Excel_Column'])
```
and then I process them with BERTopic using the command .fit_transform()

Here is my code so far:
```
vectorizer_model = CountVectorizer(ngram_range=(1,6), stop_words='english', min_df=5)

sentence_model = SentenceTransformer("all-mpnet-base-v2")

representation_model = KeyBERTInspired(top_n_words=15, nr_repr_docs=12)

ctfidf_model = ClassTfidfTransformer(bm25_weighting=True, reduce_frequent_words=False)

umap_model_2d = UMAP(n_neighbors=16, n_components=5, metric='cosine', min_dist=0.0, random_state=42, low_memory=False)

hdbscan_model_2d = HDBSCAN(min_cluster_size=15, min_samples=5, metric='euclidean', cluster_selection_method='eom', gen_min_span_tree=True, prediction_data=True)

bertopic_model = BERTopic(
                    vectorizer_model=vectorizer_model,
                    embedding_model=sentence_model,
                    umap_model = umap_model_2d,
                    hdbscan_model = hdbscan_model_2d,
                    representation_model=representation_model,
                    ctfidf_model=ctfidf_model,
                    nr_topics='auto',
                    language="english",
                    low_memory=False,
                    verbose=True, 
                    calculate_probabilities=True, 
                    top_n_words=15,
                    n_gram_range=(1,6))

embeddings = sentence_model.encode(docs, show_progress_bar=True)

topics, probs = bertopic_model.fit_transform(docs, embeddings)
```

Next as written in an answer above, I export the results obtained with the ```bertopic_model.get.documents.info()``` command to excel so that I can inspect in detail the documents contained in each topic

**Comment by MaartenGr** [comment]
There are two things that immediately pop out. First, is this:

> The data are without stopwords and punctuation, are in lower case and have been tokenized and lemmatised using SpaCy. 

These steps generally result in worse performance since Transformer-based models expect the data not to be preprocessed that much. So I would generally advise not preprocessing the model and only removing things that are completely not related to the text, like html-tags. In other words, trying it out without preprocessing might be worthwhile. 

> nr_topics='auto'

It might be worthwhile to read a bit more about this option in the document but what it does is aggregate the topics after they have been created. This means that you will likely end up with more general and abstract topics. Removing this entirely and instead focusing on `min_cluster_size` instead is likely to result in more fine-grained topics.

**Comment by MaartenGr** [comment]
> However, this technique does not seem to work. Is there a way to do this in BERTopic?

You might be able to do this with either [semi-supervised BERTopic](https://maartengr.github.io/BERTopic/getting_started/semisupervised/semisupervised.html) or [manual BERTopic](https://maartengr.github.io/BERTopic/getting_started/manual/manual.html). With semi-supervised, you are essentially providing the model with some labels in the hope they get nudged towards their respective classes. With manual, it might make sense to update the topics labels afterwards and then feed them into BERTopic. In other words, you can manually split the "Password" topics up into level_1, 2, etc. yourself and use those labels together with all other labels in manual BERTopic.

**Comment by ruVdi** [comment]
Thank you for your help, I will try to use a semi-supervised BERTopic model. Regarding the original problem in my post, I can consider it solved and if you want you can close the ticket.
